kubectl-plugin/config/migrate.go

```diff
@@ -117,23 +130,46 @@ func (o *migrateOption) getDevOpsPassword() (password string, err error) {
 	return
 }
 
-func updateAuthWithObj(yamlf string, dataMap map[string]interface{}) string {
-	mapData := make(map[string]interface{})
-	if err := yaml.Unmarshal([]byte(yamlf), mapData); err == nil {
-		var obj interface{}
-		var ok bool
-		var mapObj map[string]interface{}
-		if obj, ok = mapData["devops"]; ok {
-			mapObj = obj.(map[string]interface{})
-		} else {
-			mapObj = make(map[string]interface{})
-			mapData["devops"] = mapObj
-		}
+func (o *migrateOption) getKubeSphereConfig(configMapName, namespace string) (map[string]interface{}, *unstructured.Unstructured, error) {
+	kubesphereConfigCM, err := o.client.Resource((types.GetConfigMapSchema())).
+		Namespace(namespace).
+		Get(context.Background(), configMapName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot found ConfigMap %s/%s, %v", namespace, configMapName, err)
+	}
+	kubesphereConfigYAMLString, found, err := unstructured.NestedString(kubesphereConfigCM.UnstructuredContent(), "data", "kubesphere.yaml")
+	if err != nil {
+		return nil, nil, err
+	}
+	if !found {
+		return nil, nil, fmt.Errorf("cannot found 'kubesphere.yaml' configuration in ConfigMap %s/%s", namespace, configMapName)
+	}
+	kubesphereConfig := make(map[string]interface{})
+	if err := yaml.Unmarshal([]byte(kubesphereConfigYAMLString), kubesphereConfig); err != nil {
+		return nil, nil, err
+	}
+	return kubesphereConfig, kubesphereConfigCM, nil
+}
 
-		for key, val := range dataMap {
-			mapObj[key] = val
+// mergeMap merges patch map into main map.
+// This function has a constriant that map types must be `map[string]interface{}`, including types of map value or map value's value.
```

Comment:
Hi @LinuxSuRen , please read this comment line about the constraint of this function. You will be able to understand why test code is like that.

Reply:
Eventually, I might be able to understand this function. But there's no way to guess what do you want to do according to the name of this function. All potential contributors need to read all your comments and codes to understand them. This is not acceptable.

Reply:
Done, please check https://github.com/kubesphere-sigs/ks/pull/245/commits/6798e1427f65159770879d9c2049311c3e603e89.